### DISCUSSION TITLE
How to enable and configure the previewgenerator app

### DISCUSSION THREAD

**Discussion by szaimen** [nextcloud/all-in-one#2810] [Wiki]
Here is how to enable and configure the previewgenerator app.

0. If not already done, install the previewgenerator app and run `sudo docker exec --user www-data -it nextcloud-aio-nextcloud php occ preview:generate-all` to start the initial preview generation. Only continue after this step is done
1. Edit your cronjobs with `sudo crontab -u root -e` (and choose your editor of choice if not already done. I'd recommend nano)
2. add a new line to this file if not already present: `*/10 * * * * docker ps --format "{{.Names}}" | grep -q "^nextcloud-aio-nextcloud$" && docker exec --user www-data nextcloud-aio-nextcloud php occ preview:pre-generate` which will run the command every 10 minutes.
3. Now save and close the file (when using nano are the shortcouts for this `Ctrl + o` -> `Enter` and close the editor with `Ctrl + x`)
4. And you are done.

I hope it helps!

**Comment by ebildebil** [comment]
Thank you!

**Comment by BobCui20** [comment]
Awesome !

**Comment by ShapeShifter499** [comment]
I implemented this with systemd.

Instead of the crontab entries, I added the following two files and enabled the timer.

>/etc/systemd/system/nextcloud-preview-generator.service
>```
>[Unit]
>Description=Run Nextcloud Preview Generator
>
>[Service]
>Type=oneshot
>ExecStart=/bin/bash -c "/usr/bin/docker ps --format "{{.Names}}" | /usr/bin/grep -q "^nextcloud-aio-nextcloud$" && /usr/bin/docker exec --user www-data nextcloud-aio-nextcloud php occ preview:pre-generate"
>```

>/etc/systemd/system/nextcloud-preview-generator.timer
>```
>[Unit]
>Description=Run Nextcloud Preview Generator every 10 minutes
>
>[Timer]
>OnCalendar=*:0/10
>Persistent=true
>
>[Install]
>WantedBy=timers.target
>```

**Reply by liquidat** [reply]
Great, thanks! This still has to be enabled, right?
```
sudo systemctl enable --now nextcloud-preview-generator.timer
```

**Reply by ShapeShifter499** [reply]
Yes,  I'm sorry if that wasn't clear.

**Comment by gfro84** [comment]
Very helpful, I was trying to run a cron job from within the docker. Now the host system runs the cron job which tell docker to do the image update.

**Comment by JartanFTW** [comment]
@szaimen Using the host's crontab would result in breakage if the AIO containers were moved to a different host. How should I go about adding the cronjob within the nextcloud docker container, rather than the host?

This problem was also mentioned here: https://github.com/nextcloud/all-in-one/discussions/542#discussioncomment-8051923

I may have already achieved this, however, the `preview:pre-generate` command gives no output or logs (that I can find), so I don't know how to tell if it's running as expected. If you can tell me how I might check this, and if I've done it successfully, I'll post some instructions on how I set it up.

Furthermore, based on the [Background Job Configuration docs page](https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/background_jobs_configuration.html), I believe that previewgenerator should be registering actions with cron.php rather than demanding users to setup cronjobs? I could be wrong here:
> Nextcloud apps register actions with cron.php automatically to take care of typical housekeeping operations, such as garbage collecting of temporary files or checking for newly updated files using filescan() for externally mounted file systems.
